[libc][docs] codify Policy on Assembler Sources (#88185)
llvm/llvm-project: libc/docs/dev/code_style.rst

compiler versions`__ (and newer).

.. __: https://libc.llvm.org/compiler_support.html#minimum-supported-versions

> Policy on Assembler sources

Reviewer: Pet peeve: assembly (language), not assembler (just because the latter is in the dictionary, used by some people and what -x accepts doesn't mean it's a sensible name for it; assembler is really the tool not the language)

Author: 257fbfaeb0d3

===========================

Coding in high level languages such as C++ provides benefits relative to low
level languages like Assembler, such as:

* Improved safety
* Instrumentation

  * Code coverage
  * Profile collection
* Sanitization
* Debug info

> While its not impossible to have Assembler code that correctly provides all of

Reviewer: ```suggestion
While it's not impossible to have Assembler code that correctly provides all of
```

Author: 257fbfaeb0d3

> While its not impossible to have Assembler code that correctly provides all of

Reviewer 2: it's

the above, we do not wish to maintain such Assembler sources in llvm-libc.

> That said, there a few functions provided by llvm-libc that are more difficult
> to implement or maintain in C++ than Assembler. We do use inline or out-of-line

Reviewer: ```suggestion
That said, there a few functions provided by llvm-libc that are impossible
to reliably implement in C++. We do use inline or out-of-line
```

Reviewer 3: Maybe add, "...to reliably implement in C++ for all compilers supported for building the libc."

Reviewer 2: yeah, that's a point worth making explicit --- functions like this were hard enough for bionic to deal with, and we only had to support gcc _or_ clang at any given time. just the different versions of the _same_ compiler were painful enough, without having to worry about two completely different compilers. (and we did junk several such functions as we transitioned from gcc to clang!)

Author: I've reworded this in c06aa9d3d5ee to:

> That said, there a few functions provided by llvm-libc that are impossible
to reliably implement in C++ for all compilers supported for building
llvm-libc.

Assembly in an intentionally minimal set of places; typically places where the
stack or individual register state must be manipulated very carefully for
correctness.

> Contributions adding Assembly for performance are not welcome. Contributors

Reviewer 3: "...are not welcome.  If you feel you can write better assembly than the compiler can generate, file a bug against LLVM, as we should instead be teaching the compiler to generate the better code."

Author: updated the wording here in https://github.com/llvm/llvm-project/commit/0749f082f33b30fb3fdd5f8888177c4ade17dd6b

> Ideally, bugs should be filed against compiler vendors, and links to those bug reports should appear in commit messages or comments that seek to add Assembly to llvm-libc.

PTAL

> Contributions adding Assembler for performance are not welcome. Contributors

Reviewer 2: you're not carving out an exception for the string/memory functions? or just punting that down the road until commercial users show up?

explicitly mention that anyone wanting to go that route should consider whether they can do it via a builtin? (like all the math functions that turn into a single instruction on modern architectures, or stuff like ffs() or abs() etc?)

Reviewer 3: We already have commercial users of the libc's string/memory functions (i.e. Google production).

Reviewer 2: aye, but google3 is a bit of a monoculture... i meant something more like Android, where every SoC vendor will want workarounds for their SoC's (and SoCs') specific characteristics. commercial in the sense of "where people sell products based on benchmark scores", for better and -- mostly -- worse.

Reviewer 3: I added a new suggested edit, in general if you can write "better" assembly than the compiler can generate, we should teach the compiler to generate the better code.

Reviewer 4: +1 on teaching compilers to generate better code.

OTOH, as a library, we do have to support older compiler versions than the latest ones, so maybe in case of performance-critical, we can consider case-by-case with:
- upstream compilers' issues / PRs to improve the generated codes,
- deprecation timeline / plan when the minimum supported compilers' versions get updated,
- something else?

Reviewer 3: I go back and forth on having such a carve-out now, or waiting to see what example actually crops up.  And part of me is OK with a philosophy of "to get the best code gen, you have to use the latest compiler version" though I know not everyone agrees with that.

Reviewer 2: yeah, that's easier for Android where we explicitly only support one version of one compiler at any given time.

in my experience though you're unlikely to have people _actually_ take you up on this. certainly not the same people who'll want to give you -- naming no names -- "a binary blob in ASCII form, containing the output of our proprietary compiler on source that may or may not be the same as your source", who are the usual people _i_ have to tell to go away and improve llvm instead :-) (this is typically for libm stuff.)

the libc stuff is always the string/memory routines, and there it's going to be stuff like https://github.com/ARM-software/optimized-routines/issues/68 that you're not going to want anyway. luckily it's been quite quiet for us for a decade, as arm64 was effectively single-source, arm32 was dead, and x86/x86-64 weren't commercially significant (though we do have all the usual `s*e*[2-4]` and `avx\d+` variants anyway). personally i'd be very interested to see how many of those you can make go away with function multiversioning. (but probably never the non-temporal stuff, and convincing people who're paid to make benchmarks go up that making benchmarks go up is useless, well, "It is difficult to get a man to understand something when his salary depends upon his not understanding it".)

Author: Ah! I didn't realize we _did_ have inline asm for these kind of optimizations.  @gchatelet is attending EuroLLVM this week, but maybe next week he can catch up on this thread and help us draw the distinction of precisely where we want to draw the line.

Author: Thanks all for the feedback.  I've updated this in 0749f082f33b30fb3fdd5f8888177c4ade17dd6b. PTAL.

I now mention:
- > instances where a specific instruction sequence does not have a corresponding compiler builtin function today.
- > Contributions adding functions implemented purely in Assembly for performance are not welcome.
- > Ideally, bugs should be filed against compiler vendors, and links to those bug reports should appear in commit messages or comments that seek to add Assembly to llvm-libc.
- > Patches containing any amount of Assembly ideally should be approved by 2 maintainers.

compiler versions`__ (and newer).

.. __: https://libc.llvm.org/compiler_support.html#minimum-supported-versions

Policy on Assembly sources
===========================

Coding in high level languages such as C++ provides benefits relative to low
level languages like Assembly, such as:

* Improved safety
* Compile time diagnostics
* Instrumentation

  * Code coverage
  * Profile collection
* Sanitization
* Automatic generation of debug info

While it's not impossible to have Assembly code that correctly provides all of
the above, we do not wish to maintain such Assembly sources in llvm-libc.

> That said, there a few functions provided by llvm-libc that are impossible

Reviewer 5: there **are** a few

Author: thanks! fixed in a9c0ce3a1e3f
